### ISSUE TITLE
HTTP2 connections not used efficiently

### ISSUE THREAD

**Issue by kitkars** [reactor/reactor-netty#2262]
When we use HTTP2, single connection is enough for a remote host to send multiple requests. Unlike HTTP1.1, HTTP2 does not wait for the previous request to complete to send another request. 

**Server-Controller:**

```
    @GetMapping("echo")
    public Mono<String> echo(){
        return Mono.just("echo")
                   .delayElement(Duration.ofSeconds(1));
    }
```    
**Server-Config:**

```
    @Bean
    public NettyServerCustomizer customizer() throws CertificateException {
        return httpServer -> httpServer
                .protocol(HttpProtocol.H2C);
    }
```

**Client:**

```
    private HttpClient httpClient = HttpClient.create(ConnectionProvider.builder("http2").maxConnections(1).build())
                                              .protocol(HttpProtocol.H2C);
    private WebClient client = WebClient.builder()
                                        .baseUrl("http://localhost:8080/echo")
                                        .clientConnector(new ReactorClientHttpConnector(httpClient))
                                        .build();
                                        
    @Test
    public void test(){
        var start = System.currentTimeMillis();
        Flux<String> flux = Flux.range(1, 10)
                             .flatMap(i -> client.get().retrieve().bodyToMono(String.class))
                              .doFinally(s -> System.out.println("Total Time : " + (System.currentTimeMillis() - start) + " ms"));
        StepVerifier.create(flux.then())
                .verifyComplete();
    }                                    
```

_**Expected Behavior:**_
All the 10 requests should get completed in ~1 second. 

**_Actual Behavior:_**
All the 10 requests get completed in ~5 seconds. (To be honest, I have no clue why 5 seconds. if we send requests 1 after another, then it should have taken 10 seconds)

I tried with `maxConcurrentStreams(10)` - no luck. Same behavior.

**Comment by kitkars** [comment]
Thanks @violetagg. 

**My dependencies:**

![image](https://user-images.githubusercontent.com/44539604/171869149-831baea2-0ad4-4ac0-8489-20f6233b91f9.png)

**Server-Controller:**
```
    @GetMapping("echo/{id}")
    public Mono<String> echo(@PathVariable int id) {
        return Mono.just("echo " + id)
                   .doFirst(() -> System.out.println("Received : " + id))
                   .doFinally(s -> System.out.println("Responded : " + id))
                   .delayElement(Duration.ofSeconds(1));
    }
```
**Server Config:**
```
    @Bean
    public NettyServerCustomizer customizer() {
        return httpServer -> httpServer
                .protocol(HttpProtocol.H2C)
                .http2Settings(b -> b.maxConcurrentStreams(100));
    }
```

**Client:**
```
    private HttpClient httpClient = HttpClient.create(ConnectionProvider.builder("http2")
                                                                        .maxConnections(1).build())
                                              .protocol(HttpProtocol.H2C)
                                              .http2Settings(b -> b.maxConcurrentStreams(10));
    private WebClient client = WebClient.builder()
                                        .baseUrl("http://localhost:8080/echo/")
                                        .clientConnector(new ReactorClientHttpConnector(httpClient))
                                        .build();

    @Test
    public void test() {
        Flux<String> flux = Flux.range(1, 10)
                                .flatMap(i -> client.get()
                                                    .uri("{id}", i)
                                                    .retrieve()
                                                    .bodyToMono(String.class)
                                                    .doOnNext(r -> System.out.println(LocalDateTime.now() + ": Received : " + r)));
        StepVerifier.create(flux.then())
                    .verifyComplete();
    }
```

What I had observed:
I still see similar issues. If you look at the server log, I am not sure why we close channel for every 2 responses. (from where this 2 comes from?)

**_Server side log:_** 
```
Received : 2
Responded : 2
2022-06-03 07:59:29.815 DEBUG 36472 --- [ctor-http-nio-2] r.n.http.server.HttpServerOperations     : [f0e2587b/2-1, L:/127.0.0.1:8080 ! R:/127.0.0.1:53362] Last HTTP packet was sent, terminating the channel
...
...
Received : 4
Responded : 4
2022-06-03 07:59:30.883 DEBUG 36472 --- [ctor-http-nio-2] r.n.http.server.HttpServerOperations     : [f0e2587b/3-1, L:/127.0.0.1:8080 ! R:/127.0.0.1:53362] Last HTTP packet was sent, terminating the channel
```
**_Client side log:_**
Even though we can have 100 active streams, l am able to achieve only 2 as the server seems to keep closing.
```
{uri=http://localhost:8080/echo/2, method=GET}
2022-06-03 07:59:28.600 DEBUG 36498 --- [ctor-http-nio-2] r.n.http.client.Http2ConnectionProvider  : [a5af7766/2-1, L:/127.0.0.1:53362 - R:localhost/127.0.0.1:8080] Stream opened, now: 2 active streams and 100 max active streams.
```

**_Attached the log files for your reference._**

[netty-http2-client-log.txt](https://github.com/reactor/reactor-netty/files/8832881/netty-http2-client-log.txt)
[netty-http2-server-log.txt](https://github.com/reactor/reactor-netty/files/8832883/netty-http2-server-log.txt)

**Comment by violetagg** [comment]
@kitkars I updated the PR #2257 and it is now in its final state.

Thanks for the example that you provided. However I need to clarify some things:

The API below configures the HTTP/2 initial settings. 
`.http2Settings(b -> b.maxConcurrentStreams(N));` this configures `SETTINGS_MAX_CONCURRENT_STREAMS `. According to the specification

https://datatracker.ietf.org/doc/html/rfc7540#section-5.1.2
```
A peer can limit the number of concurrently active streams using the
   SETTINGS_MAX_CONCURRENT_STREAMS parameter (see [Section 6.5.2](https://datatracker.ietf.org/doc/html/rfc7540#section-6.5.2)) within
   a SETTINGS frame.  The maximum concurrent streams setting is specific
   to each endpoint and applies only to the peer that receives the
   setting.  That is, clients specify the maximum number of concurrent
   streams the server can initiate, and servers specify the maximum
   number of concurrent streams the client can initiate.
```

On the server you have

`.http2Settings(b -> b.maxConcurrentStreams(100));` which means that the client can open maximum 100 concurrent streams

On the client you have 

`.http2Settings(b -> b.maxConcurrentStreams(10));` which means that the server (server push) can open maximum 10 concurrent streams. However Reactor Netty does not support server push.

If your intention was to further limit the streams that the client can open, then with the mentioned PR above we are introducing a new API so that you can do this

```
reactor.netty.http.client.Http2AllocationStrategy
...
/**
 * Configures the maximum number of the concurrent streams that can be opened to the remote peer.
 * When evaluating how many streams can be opened to the remote peer,
 * the minimum of this configuration and the remote peer configuration is taken.
 * Default to {@code -1} - use always the remote peer configuration.
 *
 * @param maxConcurrentStreams the maximum number of the concurrent streams that can be opened to the remote peer
 * @return {@code this}
 */
Builder maxConcurrentStreams(long maxConcurrentStreams);
...
```

Or the `ConnectionProvider` configuration should look like this

```
Http2AllocationStrategy strategy = Http2AllocationStrategy.builder().maxConnections(1).maxConcurrentStreams(10).build();
ConnectionProvider.builder("http2")
		.allocationStrategy(strategy)
		.build()
```

Can you please test with the final changes and can you please clarify what was you original intention with `.http2Settings(b -> b.maxConcurrentStreams(10));` setting on the client.

**Comment by kitkars** [comment]
@violetagg 

Your last PR update fixes all the problems I was facing. 🥇 👍 

-------

**My setup:**

-  The server takes 3 seconds to process each reuqest (using `delayElement(Duration.ofSeconds(3))`. 
- The client sets up 1 http connection with 1 server (`.maxConnections(1)`)

My expectation was whatever the requests client sends - they all should complete in more or less ~3 seconds.

I have been playing with this for sometime. Just to quickly summarize

**Scenarios:**
- Client sends 100 requests in parallel (`Flux.range(1, 100)`) - takes ~3 seconds 
- Client sends 10K requests in parallel `(Flux.range(1, 10000) & flatMap concurrency is set to 10000` ) takes ~4 seconds (pretty good)

--------
Regarding your question for my original intention behind setting `.http2Settings(b -> b.maxConcurrentStreams(10));` - I had assumed that the client could further limit the server's `maxConcurrentStreams` from its side. Not sure of the exact use case. But I felt it would be kind of client-side rate-limiting or a way of conveying client's ability to handle the response processing rate. But you seem to have got it right and handled it via new API which seems to work great. 👍  

But this is where I find it little bit confusing (from user's perspective). That is - why do we have `maxConnections` in multiple places?

```
ConnectionProvider.builder("test")
                      .maxConnections(1) 
                      .allocationStrategy(Http2AllocationStrategy
                                                                .builder()
                                                                .maxConnections(1)
                                                                .build())
                      .build()
 ```

**Comment by violetagg** [comment]
@kitkars 

> But this is where I find it little bit confusing (from user's perspective). That is - why do we have maxConnections in multiple places?

If you provide an allocation strategy then you do not need to configure `maxConnections` outside of it.

I pointed this in the javadoc but if you think we need to do more clarifications, please tell us.

```
/**
 * Set the options to use for configuring {@link ConnectionProvider} maximum connections per connection pool.
-> * This is a pre-made allocation strategy where only max connections is specified.
-> * Custom allocation strategies can be provided via {@link #allocationStrategy(AllocationStrategy)}.
 * Default to {@link #DEFAULT_POOL_MAX_CONNECTIONS}.
 *
 * @param maxConnections the maximum number of connections (per connection pool) before start pending
 * @return {@literal this}
 * @throws IllegalArgumentException if maxConnections is negative
 */
public final SPEC maxConnections(int maxConnections) {
```

```
/**
 * Limits in how many connections can be allocated and managed by the pool are driven by the
-> * provided {@link AllocationStrategy}. This is a customization escape hatch that replaces the last
-> * configured strategy, but most cases should be covered by the {@link #maxConnections()}
-> * pre-made allocation strategy.
 *
 * @param allocationStrategy the {@link AllocationStrategy} to use
 * @return {@literal this}
 * @see #maxConnections()
 * @since 1.0.20
 */
public final SPEC allocationStrategy(AllocationStrategy<?> allocationStrategy) {
```

**Comment by kitkars** [comment]
Yeah, That would help. 
Thanks @violetagg for your prompt response and hard work to make other developer's (like me) life easier.  
Please close this issue.
I can't wait to work on final release version.

**Comment by violetagg** [comment]
@kitkars Thanks for testing the changes. Fix #2257 is available in 1.0.20-SNAPSHOT.
